# Lab | Hypothesis Testing

**Objective**

Welcome to the Hypothesis Testing Lab, where we embark on an enlightening journey through the realm of statistical decision-making! In this laboratory, we delve into various scenarios, applying the powerful tools of hypothesis testing to scrutinize and interpret data.

From testing the mean of a single sample (One Sample T-Test), to investigating differences between independent groups (Two Sample T-Test), and exploring relationships within dependent samples (Paired Sample T-Test), our exploration knows no bounds. Furthermore, we'll venture into the realm of Analysis of Variance (ANOVA), unraveling the complexities of comparing means across multiple groups.

So, grab your statistical tools, prepare your hypotheses, and let's embark on this fascinating journey of exploration and discovery in the world of hypothesis testing!

**Challenge 1**

In this challenge, we will be working with pokemon data. The data can be found here:

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/pokemon.csv

In [35]:
#libraries
import pandas as pd
import scipy.stats as st
import numpy as np


In [36]:
df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/pokemon.csv")
df

,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
0,Bulbasaur,Grass,Poison,45,49,49,65,65,45,1,False
1,Ivysaur,Grass,Poison,60,62,63,80,80,60,1,False
2,Venusaur,Grass,Poison,80,82,83,100,100,80,1,False
3,Mega Venusaur,Grass,Poison,80,100,123,122,120,80,1,False
4,Charmander,Fire,NaN,39,52,43,60,50,65,1,False
...,...,...,...,...,...,...,...,...,...,...,...
795,Diancie,Rock,Fairy,50,100,150,100,150,50,6,True
796,Mega Diancie,Rock,Fairy,50,160,110,160,110,110,6,True
797,Hoopa Confined,Psychic,Ghost,80,110,60,150,130,70,6,True
798,Hoopa Unbound,Psychic,Dark,80,160,60,170,130,80,6,True


- We posit that Pokemons of type Dragon have, on average, more HP stats than Grass. Choose the propper test and, with 5% significance, comment your findings.

Hipotesis - Dragon vs Grass HP
- H0: El HP promedio de Dragon = HP promedio de Grass (no hay diferencia)
- H1: El HP promedio de Dragon > HP promedio de Grass
- Nivel de significancia o alfa α = 0.05

In [37]:
print(df.columns.tolist())

['Name', 'Type 1', 'Type 2', 'HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed', 'Generation', 'Legendary']


In [38]:
#code here

# separamos grupos
dragon = df[(df['Type 1'] == 'Dragon') | (df['Type 2'] == 'Dragon')]['HP']
grass  = df[(df['Type 1'] == 'Grass')  | (df['Type 2'] == 'Grass')]['HP']

print(f"HP medio Dragon: {dragon.mean().round(2)}")
print(f"HP medio Grass:  {grass.mean().round(2)}")

HP medio Dragon: 82.9
HP medio Grass:  66.05


In [39]:
# Two Sample T-Test
stat, p_value = st.ttest_ind(dragon, grass, alternative= 'greater')

print(f"p-value: {p_value:.4f}")

if p_value < 0.05:
    print("Rechazamos H0: los Dragon tienen significativamente más HP que los Grass.")
else:
    print("No rechazamos H0: no hay evidencia suficiente")

p-value: 0.0000
Rechazamos H0: los Dragon tienen significativamente más HP que los Grass.


Se aplicó Two Sample T-Test para comparar el HP medio de los Pokémon de tipo Dragon (82.9) con los de tipo Grass (66.05). 

El test devuelve un p-value de 0.0000, inferior al nivel de significancia del 5%. Por tanto, se rechaza la hipótesis nula y concluimos que los Pokémon Dragon tienen, en promedio, un HP significativamente mayor que los Grass.

- We posit that Legendary Pokemons have different stats (HP, Attack, Defense, Sp.Atk, Sp.Def, Speed) when comparing with Non-Legendary. Choose the propper test and, with 5% significance, comment your findings.


**Hipotesis  Lengendarios vs no Lengendarios**

HO : stats_Lengendario = stats_no_Lengendario

H1 : stats_Lengendario ≠ stats_no_Lengendario

In [40]:
#code here

legendario     = df[df['Legendary'] == True]
no_legendario  = df[df['Legendary'] == False]

print(f"Legendarios:     {len(legendario)}")
print(f"No legendarios:  {len(no_legendario)}")

Legendarios:     65
No legendarios:  735


In [41]:
stats = ['HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed']

for stat in stats: # para recorrer la lista del stats
    stat_leg    = legendario[stat]
    stat_no_leg = no_legendario[stat]
    
    stat_t, p_value = st.ttest_ind(stat_leg, stat_no_leg)
    
    conclusion = "Rechazamos H0" if p_value < 0.05 else "No rechazamos H0"

    print(f"{stat} , p-value: {p_value:.4f} , {conclusion}")

HP , p-value: 0.0000 , Rechazamos H0
Attack , p-value: 0.0000 , Rechazamos H0
Defense , p-value: 0.0000 , Rechazamos H0
Sp. Atk , p-value: 0.0000 , Rechazamos H0
Sp. Def , p-value: 0.0000 , Rechazamos H0
Speed , p-value: 0.0000 , Rechazamos H0


- Los Pokémon Legendarios tienen stats significativamente diferentes a los no legendarios en todas las categorías — HP, Attack, Defense, Sp. Atk, Sp. Def y Speed.

- Por eso, no es casualidad que  los legendarios son los más poderosos del juego.

Two Sample T-Test para cada stat comparando Pokémon Legendarios vs No Legendarios. En todos los casos el p-value es inferior a 0.05, por lo que rechazamos H0 y concluimos que los Pokémon Legendarios tienen stats significativamente diferentes a los no legendarios y por tanto hay evidencia para apoyar la hipotesis alternativa h1.

**Challenge 2**

In this challenge, we will be working with california-housing data. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/california_housing.csv

In [42]:
df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/california_housing.csv")
df.head()

,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value
0,-114.31,34.19,15.0,5612.0,1283.0,1015.0,472.0,1.4936,66900.0
1,-114.47,34.40,19.0,7650.0,1901.0,1129.0,463.0,1.8200,80100.0
2,-114.56,33.69,17.0,720.0,174.0,333.0,117.0,1.6509,85700.0
3,-114.57,33.64,14.0,1501.0,337.0,515.0,226.0,3.1917,73400.0
4,-114.57,33.57,20.0,1454.0,326.0,624.0,262.0,1.9250,65500.0


**We posit that houses close to either a school or a hospital are more expensive.**

- School coordinates (-118, 34)
- Hospital coordinates (-122, 37)

We consider a house (neighborhood) to be close to a school or hospital if the distance is lower than 0.50.

Hint:
- Write a function to calculate euclidean distance from each house (neighborhood) to the school and to the hospital.
- Divide your dataset into houses close and far from either a hospital or school.
- Choose the propper test and, with 5% significance, comment your findings.
 

In [43]:
# Comprobar si las casas cerca de un colegio o hospital son más caras
#H0 : El precio medio de las casas cercanas = precio medio de las casas lejanas
#H1 : El precio medio de las casas cercanas > precio medio de las casas lejanas

# Distancia euclidiana? es la distancia en línea recta entre dos puntos

In [44]:
# 1. Calcular distancia de cada casa al colegio (-118,34) y añ hospital (-122,37)

In [45]:

df['dist_colegio']  = np.sqrt((-118 - df['longitude'])**2 + (34  - df['latitude'])**2)
df['dist_hospital'] = np.sqrt((-122 - df['longitude'])**2 + (37  - df['latitude'])**2)

df[['longitude', 'latitude', 'dist_colegio', 'dist_hospital']].head()

,longitude,latitude,dist_colegio,dist_hospital
0,-114.31,34.19,3.694888,8.187319
1,-114.47,34.40,3.552591,7.966235
2,-114.56,33.69,3.453940,8.143077
3,-114.57,33.64,3.448840,8.154416
4,-114.57,33.57,3.456848,8.183508


In [46]:
# 2. Dividimos en 2 grupos

condicion = (df['dist_colegio'] < 0.50) | (df['dist_hospital'] < 0.50)

cercanas = df[condicion]
lejanas  = df[~condicion]

print(f"Casas cercanas: {len(cercanas)}")
print(f"Casas lejanas:  {len(lejanas)}")

Casas cercanas: 6829
Casas lejanas:  10171


In [47]:
# 3. T- test

stat, p_value = st.ttest_ind(cercanas['median_house_value'], lejanas['median_house_value'], alternative='greater')

print(f"p-value: {p_value:.4f}")

if p_value < 0.05:
    print("Rechazamos H₀: las casas cercanas son significativamente más caras")
else:
    print("No rechazamos H₀: no hay evidencia suficiente de diferencia")

p-value: 0.0000
Rechazamos H₀: las casas cercanas son significativamente más caras


In [48]:
print(f"Precio medio cercanas: ${cercanas['median_house_value'].mean():,.0f}")
print(f"Precio medio lejanas:  ${lejanas['median_house_value'].mean():,.0f}")

Precio medio cercanas: $246,952
Precio medio lejanas:  $180,678


Tras calcular la distancia euclidiana de cada casa al colegio y al hospital, dividimos el dataset en dos grupos:

- Casas cercanas (distancia < 0.50): 6829 casas (40% del total)
- Casas lejanas (distancia ≥ 0.50): 10171 casas (60% del total)

Aplicando un One-sided Two Sample T-Test obtenemos un p-value de 0.0000, muy por debajo del nivel de significancia del 5%.

Rechazamos H0 y confirmamos que las casas cercanas a un colegio o hospital son significativamente más caras y por tanto, se puede decir que vivir cerca de estos servicios tiene un impacto en el valor de la vivienda